### PR TITLE
Map Discord roles to NextCloud groups

### PR DESCRIPTION
**Pull request by kousu**
Fixes https://github.com/zorn-v/nextcloud-social-login/issues/390

### CODE REVIEW

**Comment by kousu** [comment]
Currently this only reads Discord role IDs, not their names, which makes them very verbose in the UI:

Screenshott

I would like it to read their names. For that, I believe it needs to call [GET/guilds/[{guild.id}/roles](https://discord.com/developers/docs/resources/guild#get-guild-roles), but for _that_ I fear that it probably requires a. creating a Bot User, b. inviting that Bot User to the target guilds, c. giving the Bot User's token to this plugin d. using the Bot User's token to call that API, all of which is quite a bit more complicated than the status quo.

It might be good enough to just impose a group mapping like @pktiuk suggested; it'll be annoying for the admin to set it up the first time and every time they add a new role ((I have no idea how to find out the role IDs from the Discord UI, they just don't appear anywhere)), but it'll be stable at least.

I think, also, this patch belongs better in [HybridAuth](https://github.com/hybridauth/hybridauth). But I tried patching the vendored copy of it under 3rdparty and couldn't get it to work, until I learned from https://github.com/zorn-v/nextcloud-social-login/commit/ef5e24c192929d7fa8d7260f1a8dab278d7d09d2 that lib/Service/ProviderService.php overrides a lot of its settings so that patching it there wouldn't work. I still think that's a good idea, but it will require some coordination -- and like, say, not everyone is going to want to use it, which means the `guilds` and `guilds.members.read` scopes need to be optional.

**Comment by kousu** [comment]
I've got group mapping protoyped. But I didn't take the time to build a whole UI for it, so so far all I have is this patch, which I'm not going to attach to the branch because it's too specific to my use case, but I want to record it here anyway because it shows what needs doing:

```diff
diff --git a/lib/Service/ProviderService.php b/lib/Service/ProviderService.php
index 5e737f3..011332e 100644
--- a/lib/Service/ProviderService.php
+++ b/lib/Service/ProviderService.php
@@ -423,6 +423,15 @@ class ProviderService
                 // TODO: /member returns roles as their ID; to get their name requires an extra API call
                 //       (and perhaps extra permissions?)
             }
+
+           // HACK: because there's no UI for mapped groups under the Discord connector like there is under OIDC or Discourse
+           //       code in our mapping here
+           $profile->data['group_mapping'] = array(
+                   '5916019825232292399' => "admin",
+                   '3908310574390159257' => "Stars",
+                   '1647820633687053577' => "Coordinators",
+
+           );
         }
 
         if (!empty($config['logout_url'])) {
```

This works! I can assign/unassign people roles in Discord, and the next time they log in (which I can force with NextCloud's "Wipe all devices" option in the /settings/users page) their groups will be updated. And with "Restrict login for users without mapped groups" turned on this means that only the roles listed there are allowed to log in, giving me some peace of mind.

Some important points:

* the keys in "group_mapping" are the strings as given by Discord; they're **not** "discord-XXXXXXXX" which
* nextcloud-social-login still _creates_ groups named "discord-XXXXXXXX"; I wish I could turn that off
* the values are the names of the NextCloud groups; and you have to be careful here, these are the **original** names of the NextCloud groups; if you "rename" a NextCloud group it doesn't actually rename it, it just adds a label overtop that masks the internal name

**Comment by zorn-v** [comment]
> the keys in "group_mapping" are the strings as given by Discord; they're not "discord-XXXXXXXX" which

Yep, mapping assumes "exact foreign group -> exact NC group".

> nextcloud-social-login still creates groups named "discord-XXXXXXXX"; I wish I could turn that off

Just disable `auto_create_groups` option

> the values are the names of the NextCloud groups; and you have to be careful here, these are the original names of the NextCloud groups; if you "rename" a NextCloud group it doesn't actually rename it, it just adds a label overtop that masks the internal name

There is `gid` and `displayname` for groups. Long time `displayname` just not used. Seems it is changed.

**Comment by pktiuk** [comment]
@kousu   
Maybe I will wait with this testing until you will apply suggestions of zorn-v

**Comment by zorn-v** [comment]
@pktiuk I did not add any useful information (almost), soo )

**Comment by kousu** [comment]
> Just disable `auto_create_groups` option

This is useful! And obvious in retrospect.

**Comment by kousu** [comment]
@pktiuk 

>  Maybe I will wait with this testing until you will apply suggestions of zorn-v

There's no code changes to make from this yet :face_with_spiral_eyes:  I haven't added UI to set up the group mappings yet. For my team, I edited `group_mapping` in directly into the code just to see if I understood how it worked, but that's not part of the branch you can use.

**Comment by kousu** [comment]
Looking ahead, I'm stumped on how to do the workflow for adding new mapped groups. With "Custom Discourse" the UI looks like this empty free-form box:

Screenshot 2023-02-17 at 09-50-39 Social login - Administration settings - Yielding Opportunities for Terpsichore's Art

Which is maybe fine for Discourse where the group names are obvious, but with Discord the group names are ID numbers, and as far as I can, the only way to read roles are to use [`GET /users/@me/guilds/[{guild.id}`](https://discord.com/developers/docs/resources/guild#guild-object)/member](https://discord.com/developers/docs/resources/user#get-current-user-guild-member), or [`GET/guilds/[{guild.id}](https://discord.com/developers/docs/resources/guild#guild-object)/roles`](https://discord.com/developers/docs/resources/guild#get-guild-roles). The former works with the plugin's current OAuth2 setup, but the latter, I think, requires additionally [creating a bot user](https://discord.com/developers/docs/topics/oauth2#bot-vs-user-accounts) and inviting it to each guild you might want to use it with. Which just seems like it greatly complicates the setup process.

**Comment by FabiChan99** [comment]
Yeah, I want this too

**Comment by kousu** [comment]
> When will the feature be integrated into the official version?

Unfortunately there's no timeline :( . It has no UI yet and what I do have written needs a lot more testing. If you want to speed it along, cheer me on, help me test, and if you want to learn a bit of PHP or Angular, read https://nextcloud.com/developer/ and give me a hand getting the code over the finish line.

**Comment by kousu** [comment]
> Obviously I would like it to read their names instead.

I bet https://discord.com/blog/usernames is going to either simplify or complexify this whole bit.

**Comment by FabiChan99** [comment]
> > When will the feature be integrated into the official version?
> 
> Unfortunately there's no timeline :( . It has no UI yet and what I do have written needs a lot more testing. If you want to speed it along, cheer me on, help me test, and if you want to learn a bit of PHP or Angular, read https://nextcloud.com/developer/ and give me a hand getting the code over the finish line.

I would help test too if you need. I need this for my project

**Comment by zorn-v** [comment]
> or Angular
angular sucks, learn vue :smile:

**Comment by kousu** [comment]
> Currently I have been relying on `auto_create_groups` and, tbh, tailing my logs, to find out what the Discord role IDs I should be looking for are. It's not been a good time.

@FabiChan99 taught me something! Discord has a Developer Mode under User -> Settings -> Advanced

Screenshot_20230807_135954

and with this enabled, you can get your Role IDs out by right clicking on them under Server -> Settings -> Roles

Screenshot_20230807_140120

That also gives me a good idea to at least get this to a usable state: the "Custom OIDC" and "Custom OAuth2"  backends have this "Add Group Mapping" button which expects you to know the OIDC/OAuth2 Group ID strings:

Screenshot_20230807_134645

If I can teach people how to turn on Disccord Developer Mode then I can copy this UI and have them paste in the Role IDs the same way. :tada: Once I get that working I'll undraft this and ask it to be merged.

I would really like to have a button that pulls the Discord roles from their API, but that requires bouncing through their OAuth flow so it's more complicated than I want to think about right now. That can be a follow up.

**Comment by kousu** [comment]
I've got it working!

## Testing

To try this out (from scratch, to give you a complete picture):

1. Install [`sociallogin`](https://apps.nextcloud.com/apps/sociallogin) from your instance's app manager. You need to be a NextCloud admin to do this.
2. Log in to a shell on your server (over ssh, but maybe another way if you roll that way)
3. `cd path/to/nextcloud/`; make sure that `ls -l apps/sociallogin/` shows you something
4. `mv sociallogin sociallogin.official.disabled && git clone -b discord-roles https://github.com/kousu/nextcloud-social-login sociallogin`
5. Connect Discord to NextCloud [as described in the README](https://github.com/kousu/nextcloud-social-login/blob/74def1119a3f17357b8b0bbf592f1df9a86a450c/README.md?plain=1#L108-L116). You must be an admin on your Discord server to do this.
6. Go to Discord, User -> Settings -> Advanced, and turn on Developer Mode to enable the next step

    Discord Developer Mode Screenshot
7. **For each** Discord role you want to map in
    1. Create a group in your NextCloud server
    2. Go to Discord -> Server -> Settings -> Roles -> find the role -> Copy Role ID
    
        Copy Role ID Screenshot
    3. Go to NextCloud -> Settings -> Admin Settings -> Social Login -> Discord -> Add Group Mapping. Paste in the Role ID and pick the matching NextCloud group

        Discord auth provider settings screenshot
    
    5. Click Save so that you see

        > Settings for social login successfully saved

        save

## Bugs

I figured out Vue but I'm sure I'm a noob at it and I could use @zorn-v's review to make this a lot better.

I especially struggled with the distinction between "custom" auth providers

https://github.com/zorn-v/nextcloud-social-login/blob/74def1119a3f17357b8b0bbf592f1df9a86a450c/lib/Settings/AdminSettings.php#L59

https://github.com/zorn-v/nextcloud-social-login/blob/74def1119a3f17357b8b0bbf592f1df9a86a450c/src/components/settings/provider-types.js#L4-L229

https://github.com/zorn-v/nextcloud-social-login/blob/74def1119a3f17357b8b0bbf592f1df9a86a450c/src/components/Settings.vue#L148

https://github.com/zorn-v/nextcloud-social-login/blob/74def1119a3f17357b8b0bbf592f1df9a86a450c/src/components/Settings.vue#L158


<details><summary>Custom Provider UI</summary>

Custom Provider UI Screenshot

</details>

and "providers"

https://github.com/kousu/nextcloud-social-login/blob/74def1119a3f17357b8b0bbf592f1df9a86a450c/lib/Settings/AdminSettings.php#L41-L51

<details><summary>Custom Provider UI</summary>

Provider UI Screenshot

</details>

I got it working but I especially don't like this copy-pasted code:

https://github.com/kousu/nextcloud-social-login/blob/74def1119a3f17357b8b0bbf592f1df9a86a450c/src/components/Settings.vue#L191-L201

or this:

https://github.com/kousu/nextcloud-social-login/blob/74def1119a3f17357b8b0bbf592f1df9a86a450c/src/components/Settings.vue#L123-L136

or this weird hack which is needed to make the "Add Group Mapping" button show. If I need weird hacks I would prefer if the provider-specific code could all live together:

https://github.com/kousu/nextcloud-social-login/blob/74def1119a3f17357b8b0bbf592f1df9a86a450c/lib/Settings/AdminSettings.php#L53-L57

## TODO

* [ ] Fix the CSS.

    For some reason the existence of the Add Group Mapping button widens the Discord box and breaks the layout.
* [ ] Massage the "Copy Role ID" install instructions into README.md

## Follow ups

* [ ] Add a "read available Discord roles" button in the settings instead of making people use Developer Mode

File: lib/Service/ProviderService.php [hunk 1]
```diff
@@ -79,6 +79,7 @@ class ProviderService
                 'id' => 'appid',
                 'secret' => 'secret',
             ],
+            'group_mapping' => 'groupMapping',
```
Comment (kousu) [inline lib/Service/ProviderService.php:82]: I don't understand what `$configMapping` is about. We control the schema, can't we just force all the providers to use the same one?

File: src/components/Settings.vue [hunk 2]
```diff
@@ -120,6 +120,20 @@
           <input type="text" :name="'providers['+name+'][guilds]'" v-model="provider.guilds"/>
         </label>
       </template>
+      <template v-if="provider.hasGroupMapping">
+        <button class="group-mapping-add" type="button" @click="provider.groupMapping.push({foreign: '', local: ''})">
```
Comment (kousu) [inline src/components/Settings.vue:124]: For some reason the existence of the Add Group Mapping button widens the Discord box and breaks the layout. I do not understand. I've already spent 40 minutes looking at it.

Seriously what is this BS

srs

Fix the CSS here.

File: src/components/Settings.vue [hunk 3]
```diff
@@ -165,6 +183,24 @@ export default {
...
+    // from { foreign0: local0, foreign1, local1, ... }
+    // to [ { 'foreign': foreign0, 'local': local0 }, { 'foreign': foreign1, 'local': local1 }, ... ]
+    // This is necessary for Vue to use groupMapping as a read-write data model.
+    for (var provType in data.providers)  {
+      if (data.providers[provType].hasGroupMapping) {
+        var groupMappingArr = []
+        var groupMapping = data.providers[provType].groupMapping
+        if (groupMapping) {
+          for (var foreignGroup in groupMapping) {
+            groupMappingArr.push({foreign: foreignGroup, local: groupMapping[foreignGroup]})
```
Comment (kousu) [inline src/components/Settings.vue:197]: This code is copy-pasted!! :anger:

File: src/components/Settings.vue [hunk 4]
```diff
@@ -120,6 +120,20 @@
...
+        <div v-for="(mapping, mappingIdx) in provider.groupMapping" :key="mapping">
+          <input type="text" class="foreign-group" v-model="mapping.foreign" />
+          <select class="local-group" :name="mapping.foreign ? 'providers['+name+'][groupMapping]['+mapping.foreign+']' : ''">
+            <option v-for="group in groups" :key="group" :value="group" :selected="mapping.local === group">
+              {{ group }}
+            </option>
+          </select>
+          <span class="group-mapping-remove" @click="provider.groupMapping.splice(mappingIdx, 1)">x</span>
+        </div>
+      </template>
```
Comment (kousu) [inline src/components/Settings.vue:136]: This code is also copy-pasted!!

File: lib/Service/ProviderService.php [hunk 5]
```diff
@@ -422,6 +441,7 @@ private function auth($class, array $config, $provider, $providerType = null)
         }
 
         $profile->data['default_group'] = $config['default_group'];
+        $profile->data['group_mapping'] = $config['group_mapping'];
```
Comment (kousu) [inline lib/Service/ProviderService.php:444]: Does this need to be wrapped in an `isset()`? My PHP is kind of weak.

**Review by kousu — COMMENTED** [review]
This still needs:

* [ ] Massage the "Copy Role ID" install instructions into README.md


---

I especially struggled with the distinction between "custom" auth providers

https://github.com/zorn-v/nextcloud-social-login/blob/74def1119a3f17357b8b0bbf592f1df9a86a450c/lib/Settings/AdminSettings.php#L59

https://github.com/zorn-v/nextcloud-social-login/blob/74def1119a3f17357b8b0bbf592f1df9a86a450c/src/components/settings/provider-types.js#L4-L229

https://github.com/zorn-v/nextcloud-social-login/blob/74def1119a3f17357b8b0bbf592f1df9a86a450c/src/components/Settings.vue#L148

https://github.com/zorn-v/nextcloud-social-login/blob/74def1119a3f17357b8b0bbf592f1df9a86a450c/src/components/Settings.vue#L158

which render at the top of the settings page:

<details><summary>Custom Provider UI</summary>

Custom Provider UI Screenshot

</details>

and can have many instances of themselves,

and "providers" like Discord.com, which only have one of each and have custom code to support their proprietary APIs scattered throughout the plugin

https://github.com/zorn-v/nextcloud-social-login/blob/74def1119a3f17357b8b0bbf592f1df9a86a450c/lib/Settings/AdminSettings.php#L41-L51


<details><summary>Provider UI</summary>

Provider UI Screenshot

</details>

They are both stored in the database as JSON, but in separate objects with similar but subtly different schemas -- particularly, because customs have to be nested in an Array because there can be many of them. This split causes a lot of special cases in the code. I got it working but I wonder if it could be more elegant.

File: lib/Service/ProviderService.php (hunk 1 above)
Comment (zorn-v) [reply lib/Service/ProviderService.php:82]: hybridauth is external lib

Comment (FabiChan99) [reply lib/Service/ProviderService.php:82]: It just works as it should. I think with the little Tutorial, everyone can set this up!

**Comment by zorn-v** [comment]
I'll take a look later (and maybe make some refactoring). In vacations now :smiley:

**Comment by zorn-v** [comment]
Implemented in https://github.com/zorn-v/nextcloud-social-login/commit/eb1ab8890c815691463888d9309c137898365b04

**Comment by pktiuk** [comment]
I have one question linked with implementation of this functionality.     
Are these groups assigned only after the first time login, or are they updated after every login?

**Comment by kousu** [comment]
> Are these groups assigned only after the first time login, or are they updated after every login?

They're updated every login. There's a setting to tune it, but by default it does what you want:

2023-08-30-082757_366x77_scrot

**Comment by kousu** [comment]
I updated my own today finally (replacing the one from my git branch with the one from [the app store](https://apps.nextcloud.com/apps/sociallogin/releases?platform=24#24)) and it's looking good!! Thanks for fixing the CSS on that button @zorn-v.

Screenshot 2023-09-13 at 20-48-08 Social login - Administration settings - Your Organic Techno Adventure

I also appreciate that the default providers are now all hidden by default unless in use!!

**Pull request closed without merging**